Plot with events

cslotboom opened the pull request:
1. Added event handling function that plots a node deformation over time.
2. Renamed some test folders for clarity (a lot of the additions/deletions are here)
3. fixed a bug that made it so zero length fiber sections couldn't be plotted.
4. Added a test for plotting ZL fiber sections
5. Added a docstring for _readNodeDispData

The Event plot function is a bit of an experiment - I wanted to get suggestions and see if we like it.
I'm not sure we would want to include it in the final package. Maybe there should be a place for experimental functions.

![Moving_files](https://user-images.githubusercontent.com/50510141/90625616-b8e8d280-e1ce-11ea-932f-1aad5e7cfbae.gif)

cslotboom committed:
Added Event picking function

Added event picking function
Renamed some test folders for for clarity
Added capability to plot ZLEs in fiber sections
Added a test for plotting ZLE
Added a docstring for _readNodeDispData

cslotboom committed:
Added additional argument for ZLE

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -205,7 +205,7 @@ def readODB(*argv):
 
 
-def saveFiberData2D(ModelName, LoadCaseName, eleNumber, sectionNumber, deltaT = 0.0):
+def saveFiberData2D(ModelName, LoadCaseName, eleNumber, sectionNumber, deltaT = 0.0, ZLE = True):
```
ZLE=True does not use "sectionNumber" as input (line 234). What section number should the user provide in that case? Does it make sense to make that input optional with a default = 1?

cslotboom replied:
In this case there should be no section, as for zero length elements (ZLE) there is no section! For some reason, the reorder is different for zero length elements, I made a test to confirm this one.

u-anurag replied:
I get that the recorder is different and there is no section for ZLE. But it looks like there has to be a user input for ``sectionNumber`` since it is not an optional argument. As a user, should I just put a random number there? for example,

`saveFiberData2D("Building","Pushover", 1, 1000)`

The above command will record fiber data of element 1 assuming a ZLE (default), irrespective of the sectionNumber I assign. Also, I think the default value for ZLE should be "False" so the users switch it to "True" only if they record a ZLE. 

`def saveFiberData2D(ModelName, LoadCaseName, eleNumber, sectionNumber=1, deltaT=0.0, ZLE="False")`

The uses for the above definition will be,

`saveFiberData2D("Building","Pushover", 1, ZLE="True)`   To record a ZLE element and,

`saveFiberData2D("Building","Pushover", 1)`   To record a default section (1) in a NL beam-column element.

`saveFiberData2D("Building","Pushover", 1, 4)`   To record a section at integration point 4 in a NL beam-column element.

cslotboom replied:
Yeah that is a good point, I think that makes sense on both accounts. ZLE should have equaled False.
I have made the proposed change.

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -582,8 +585,7 @@ def nodecoordsEigen(nodetag):
 	
 
-
-def plot_deformedshape(Model="none", LoadCase="none", tstep = -1, scale = 10, overlap='no', monitorEleTags = [], monitorOutFile = "none"):
```
I guess you need to update your fork to match this branch. We'll keep monitorEleTags input for now. It has been helpful for big structures for my research. I think this could help others.

cslotboom replied:
Yeah, this should be there.. Let me confirm. I might need to remove this pull request.

cslotboom replied:
This should be fixed.

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -1448,3 +1376,254 @@ def update_plot(ii):
+
+
+# =============================================================================
+# Experimental Functions
```
I like the idea of plot on click. I would want to keep it permanently. However, let's keep this "Experimental Functions" heading for now so we know this is something we should discuss a lot.

cslotboom replied:
Okay, good to know!

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -1448,3 +1376,254 @@ def update_plot(ii):
+        # print('onpick points:', points)
+
+    def plot(self):
+        """
```
can we just call plot_deformedshape() function here instead of duplicating the code? Problems will arise if we change things in the plot_deformedshape() function.

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -1448,3 +1376,254 @@ def update_plot(ii):
+# Experimental Functions
+# =============================================================================
+
+def _plot_on_click(self, event):
```
Is this needed? I see this function is present in ``class plot_deformedshape_events``

cslotboom replied:
It was not no.

u-anurag commented on openseespy-pip/openseespy/postprocessing/Get_Rendering.py:
```diff
@@ -1448,3 +1376,254 @@ def update_plot(ii):
+        self.timeSteps, self.Disp_nodeArray = idbf._readNodeDispData(self.Model, self.LoadCase)
+        self.nodeArray, self.elementArray = idbf._readNodesandElements(self.Model)
+
+    def _plot_on_click(self, event):
```
I see this works great for a 2D model where it plots x-disp on clicking a node. What does it do for a 3D model? Is it possible to extend this with subplots for each dof? x-disp and y-disp for a 2D model and x, y and z-disp for a 3D model.

cslotboom replied:
It is possible to extend to 3D. If this is something we like, I will do that.

u-anurag replied:
This is a very cool function to have. I like it.

u-anurag reviewed (commented):
Here is my review. Can we extend the ``plot_on_click`` to 3D models to plot x, y and z-disp?

cslotboom committed:
Added experimental file folder. Small changes to Plot Deformation

Reduced the indentation level of plot deformation using error handling
Created a Experimental file for functions under development.

cslotboom commented:
I've made a few changes:

1. I added a experimental functions file. This is intended as a test bed for new functions/features.
1. I used error handling to reduce the indentation level of plot_model_deformation()

We should remove the experimental file before each release. 
I'd just like to have it because it's hard to track changes when they all occur in the same file.

u-anurag reviewed (commented):
Check the comment regarding the ZLE fiber recorder and let me know what you think.

cslotboom committed:
Update Get_Rendering.py
